### PR TITLE
Encode read options as part of get message

### PR DESCRIPTION
**Pull request by davecromberge**
@Licenser could you give me your opinion on these changes - is there a better way to facilitate old vs new gets?
I left the setting of defaults to the client - unless you think that it should be part of the protocol and we set them here.

### CODE REVIEW

File: eqc/tcp_proto_eqc.erl [hunk 1]
```diff
@@ -83,6 +83,12 @@ resolution() ->
 ttl() ->
     oneof([infinity, mtime()]).
 
+read_repair_opt() ->
+    oneof([0, 1, 2]).
+
+r_opt() ->
+    ?SUCHTHAT(I, int(), I >= 0 andalso (I band 16#FF) =:= I).
```
Comment (Licenser) [inline eqc/tcp_proto_eqc.erl]: `choose/2` might be better here `?SUCHTHAT` leads to 'non valid tests' that show as skipped plus is probably slower then `choose/2`.

File: src/dproto_tcp.erl [hunk 2]
```diff
@@ -25,8 +25,12 @@
 -export_type([ttl/0, bucket_info/0, tcp_message/0,
               batch_message/0, stream_message/0]).
 
+
 -type ttl() :: pos_integer() | infinity.
 
+% Read repair option: 0 = off, 1 = on, 2 = default
+-type rr_opt() :: 0 | 1 | 2.
```
Comment (Licenser) [inline src/dproto_tcp.erl]: Exposing those as atoms would be a lot better. The wire representation (number) odes not have to meet the API representation.

This would lead to more readable code.

i.e. `{rr, on}` instead of `{rr, off}`

also by decuppling representation form API it allows to change representation if required.

File: src/dproto_tcp.erl [hunk 3]
```diff
@@ -75,6 +79,13 @@
          Metric :: binary(),
          Time :: pos_integer(),
          Count :: pos_integer()} |
+        {get,
+         Bucket :: binary(),
+         Metric :: binary(),
+         Time :: pos_integer(),
+         Count :: pos_integer(),
+         RR :: rr_opt(),
+         R :: non_neg_integer()} |
```
Comment (Licenser) [inline src/dproto_tcp.erl]: I would like this to be a own type with `default` and `n` as valid options that then decode to `0` and `255` as discussed.

As above passing 'special' numbers feels wrong for an API.

**Comment by davecromberge** [comment]
@Licenser thanks for your review, I agree with all your points and have tried to incorporate your feedback.  Please have another look when you have time.

File: eqc/tcp_proto_eqc.erl [hunk 4]
```diff
@@ -83,6 +83,12 @@ resolution() ->
 ttl() ->
     oneof([infinity, mtime()]).
 
+read_repair_opt() ->
+    oneof([{rr, default}, {rr, on}, {rr, default}]).
```
Comment (Licenser) [inline eqc/tcp_proto_eqc.erl]: This lists 'default' twice and no off.

also `{rr, oneof([default, on, off])}` will read more stright forward (and would probably have made the missing off more obvious ;).

File: eqc/tcp_proto_eqc.erl [hunk 5]
```diff
@@ -83,6 +83,12 @@ resolution() ->
...
+
+r_opt() ->
+    oneof([{r, default}, {r, n}, {r, choose(1, 254)}]).
```
Comment (Licenser) [inline eqc/tcp_proto_eqc.erl]: see above

File: eqc/tcp_proto_eqc.erl [hunk 6]
```diff
@@ -108,6 +114,13 @@ get_events() ->
             {get_events, bucket(), min(S, E), max(S, E), filters()}
            ])).
 
+get_metrics() ->
+    Opts = [read_repair_opt(), r_opt()],
```
Comment (Licenser) [inline eqc/tcp_proto_eqc.erl]: this should probably be it's own generator and also test against empty, and partial lists.

File: src/dproto_tcp.erl [hunk 7]
```diff
@@ -25,8 +25,12 @@
 -export_type([ttl/0, bucket_info/0, tcp_message/0,
               batch_message/0, stream_message/0]).
 
+
 -type ttl() :: pos_integer() | infinity.
 
+-type read_opts() :: [{rr, default} | {rr, off} | {rr, on} |
+                      {r, n} | {r, default} | {r, pos_integer()}].
```
Comment (Licenser) [inline src/dproto_tcp.erl]: A extracted read_opt type might be nicer to read then a list with or's in it (I've been guilty of that myself :( ). With more complex types like this 1 alternative per line usually reads better a well.

`pos_integer()` should probably be `1..254` to exclude 255.

File: src/dproto_tcp.erl [hunk 8]
```diff
@@ -418,6 +444,14 @@ decode(<<?GET,
          Time:?TIME_SIZE/?SIZE_TYPE, Count:?COUNT_SIZE/?SIZE_TYPE>>) ->
     {get, Bucket, Metric, Time, Count};
 
+decode(<<?GET,
+         _BucketSize:?BUCKET_SS/?SIZE_TYPE, Bucket:_BucketSize/binary,
+         _MetricSize:?METRIC_SS/?SIZE_TYPE, Metric:_MetricSize/binary,
+         Time:?TIME_SIZE/?SIZE_TYPE, Count:?COUNT_SIZE/?SIZE_TYPE,
+         RR:?GET_OPT_SIZE/?SIZE_TYPE, R:?GET_OPT_SIZE/?SIZE_TYPE>>) ->
+    Opts = [decode_opt({rr, RR}), decode_opt({r, R})],
```
Comment (Licenser) [inline src/dproto_tcp.erl]: any reason not to use two functions here? They will have a smaller size each and be more readable that way?

`decode_rr(RR) -> {rr, ...}`

File: src/dproto_tcp.erl [hunk 9]
```diff
@@ -283,6 +293,22 @@ encode({get, Bucket, Metric, Time, Count}) when
...
+      %% We only want positive numbers <  32 bit
+      is_integer(Count), Count > 0, (Count band 16#FFFFFFFF) =:= Count,
+      is_list(Opts) ->
+    RROpt = proplists:get_value(rr, Opts, default),
+    ROpt = proplists:get_value(r, Opts, default),
+    <<?GET,
+      (byte_size(Bucket)):?BUCKET_SS/?SIZE_TYPE, Bucket/binary,
+      (byte_size(Metric)):?METRIC_SS/?SIZE_TYPE, Metric/binary,
+      Time:?TIME_SIZE/?SIZE_TYPE, Count:?COUNT_SIZE/?SIZE_TYPE,
+      (encode_opt({rr, RROpt})):?GET_OPT_SIZE/?SIZE_TYPE,
```
Comment (Licenser) [inline src/dproto_tcp.erl]: any reason not to use a single function here (see below).

**Review by Licenser — CHANGES_REQUESTED** [review]

File: src/dproto_tcp.erl (hunk 8 above)
Comment (davecromberge) [reply src/dproto_tcp.erl]: I did that as my first implementation then i messed it up by trying to make it generic with a map as a lookup - I agree that the former version is clearer.

File: eqc/tcp_proto_eqc.erl (hunk 4 above)
Comment (davecromberge) [reply eqc/tcp_proto_eqc.erl]: good catch, really need to get a licence for eqc

File: src/dproto_tcp.erl (hunk 7 above)
Comment (davecromberge) [reply src/dproto_tcp.erl]: That's true, I realised that once I started incorporating the changes to the dalmatiner read FSM.  I started re-encoding the same type spec and an exported type would be a lot better here.

**Comment by Licenser** [comment]
EQC fails with:
```
===== tcp_proto_eqc:prop_encode_decode_general
*******************************Failed! Reason:
{'EXIT',
    {badarg,
        [{dproto_tcp,encode,1,
             [{file,
                  "/Users/heinz/Projects/dalmatiner/libs/dproto/_build/eqc/lib/dproto/src/dproto_tcp.erl"},
              {line,313}]},
         {tcp_proto_eqc,'-prop_encode_decode_general/0-fun-2-',1,
             [{file,
                  "/Users/heinz/Projects/dalmatiner/libs/dproto/_build/eqc/lib/dproto/eqc/tcp_proto_eqc.erl"},
              {line,191}]}]}}
After 31 tests.
{get,<<"t">>,<<"qiqx">>,11758696116143161345,4049206326,
     [{rr,default},{r,default}]}
Shrinking ******(6 times)
Reason:
{'EXIT',
    {badarg,
        [{dproto_tcp,encode,1,
             [{file,
                  "/Users/heinz/Projects/dalmatiner/libs/dproto/_build/eqc/lib/dproto/src/dproto_tcp.erl"},
              {line,313}]},
         {tcp_proto_eqc,'-prop_encode_decode_general/0-fun-2-',1,
             [{file,
                  "/Users/heinz/Projects/dalmatiner/libs/dproto/_build/eqc/lib/dproto/eqc/tcp_proto_eqc.erl"},
              {line,191}]},
         {eqc_lazy_lists,lazy_safe_map,2,
             [{file,
                  "/Users/thomas/Quviq/Development/eqc-dev/eqc-project/ebin/../../eqc-project/src/eqc_lazy_lists.erl"},
              {line,38}]}]}}
{get,<<"a">>,<<"a">>,1,1,[]}
```

**Comment by Licenser** [comment]
EQC is still angry:

```erlang
%% ===== tcp_proto_eqc:prop_encode_decode_general
%% ******************************Failed! After 30 tests.
{get,<<"m">>,<<"npc">>,10012436359065698305,2925196774,
     [{r,204},{rr,off},{r,146}]}
{get,<<"m">>,<<"npc">>,10012436359065698305,2925196774,
     [{r,204},{rr,off},{r,146}]} -> <<2,1,109,0,3,110,112,99,138,243,81,210,6,
                                      81,0,1,174,90,245,230,0,204>> -> {get,
                                                                        <<"m">>,
                                                                        <<"npc">>,
                                                                        10012436359065698305,
                                                                        2925196774,
                                                                        [{r,
                                                                          204},
                                                                         {rr,
                                                                          off}]}
%% Shrinking ******(6 times)
{get,<<"a">>,<<"a">>,1,1,[]}
{get,<<"a">>,<<"a">>,1,1,[]} -> <<2,1,97,0,1,97,0,0,0,0,0,0,0,1,0,0,0,1,2,0>> -> {get,
                                                                                  <<"a">>,
                                                                                  <<"a">>,
                                                                                  1,
                                                                                  1,
                                                                                  [{r,
                                                                                    default},
                                                                                   {rr,
                                                                                    default}]}
```
